File: web/__tests__/mark-href.test.ts
/**
 * Tests for `markHref` — the ATTRS-FIRST link-mark href resolver that keeps the
 * web reader in parity with the Go (`internal/pdrender/inline.go`) and Elixir
 * (`inline.ex`) renderers. A non-empty `attrs.href` (ProseMirror shape) wins,
 * else the top-level `href`; anything else is `undefined`. Precedence and edge
 * cases mirror `internal/pdrender/markhref_test.go`.
 *
 * Run: `pnpm test` (or `cd web && node --test __tests__/mark-href.test.ts`).
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import { markHref } from "../lib/mark-href.ts";

test("resolves nested attrs.href", () => {
  assert.equal(markHref({ attrs: { href: "https://x" } }), "https://x");
});

test("resolves top-level href when no attrs", () => {
  assert.equal(markHref({ href: "https://y" }), "https://y");
});

test("attrs.href wins when both present", () => {
  assert.equal(
    markHref({ attrs: { href: "https://attrs" }, href: "https://top" }),
    "https://attrs",
  );
});

test("empty attrs.href falls through to top-level", () => {
  assert.equal(
    markHref({ attrs: { href: "" }, href: "https://z" }),
    "https://z",
  );
});

test("attrs present without href key falls through to top-level", () => {
  assert.equal(
    markHref({ attrs: { title: "t" }, href: "https://w" }),
    "https://w",
  );
});

test("returns undefined for string mark, null, and number", () => {
  assert.equal(markHref("not a map"), undefined);
  assert.equal(markHref(null), undefined);
  assert.equal(markHref(42), undefined);
});

test("returns undefined when neither href is present", () => {
  assert.equal(markHref({ foo: "bar" }), undefined);
});

File: internal/pdrender/markhref_test.go
package pdrender

import "testing"

// markHref resolves a link mark's href: nested attrs.href wins, else a
// top-level href, else "". It feeds the OSC-8 hyperlink path (sanitizeURL),
// so a wrong lookup would drop or mis-target terminal links.
func TestMarkHref(t *testing.T) {
	cases := []struct {
		name string
		in   any
		want string
	}{
		{"nested attrs.href", map[string]any{"attrs": map[string]any{"href": "https://x"}}, "https://x"},
		{"top-level href (no attrs)", map[string]any{"href": "https://y"}, "https://y"},
		{"attrs present but empty href → top-level fallback", map[string]any{"attrs": map[string]any{"href": ""}, "href": "https://z"}, "https://z"},
		{"attrs present, no href key → top-level fallback", map[string]any{"attrs": map[string]any{"title": "t"}, "href": "https://w"}, "https://w"},
		{"neither → empty", map[string]any{"foo": "bar"}, ""},
		{"non-map input → empty", "not a map", ""},
		{"nil → empty", nil, ""},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := markHref(tc.in); got != tc.want {
				t.Errorf("markHref(%v) = %q, want %q", tc.in, got, tc.want)
			}
		})
	}
}
